File: solution/ts/week05-complexity/exercise2/src/birthdayCommand.ts
import {Command} from "./command";

export class BirthdayCommand implements Command {
    private readonly today: () => Date;
    private readonly birthday: Date;

    constructor(today: () => Date, birthday: Date) {
        this.today = today;
        this.birthday = birthday;
    }

    executeAndDisplayResult(): string {
        let daysBetween = this.daysBetween(this.today(), this.birthday);

        if (daysBetween < 0) {
            const nextYearBirthday = new Date(this.birthday);
            nextYearBirthday.setFullYear(this.today().getFullYear() + 1);
            daysBetween = this.daysBetween(this.today(), nextYearBirthday);
        }
        return `Still ${daysBetween} days left!`;
    }

    private daysBetween(today: Date, other: Date) {
        return Math.ceil((other.getTime() - today.getTime()) / (1000 * 60 * 60 * 24));
    }
}

File: solution/c#/week05-complexity/Exercise2CommandProcessor.Tests/BirthdayCommand.cs
namespace Command.Tests;

public class BirthdayCommand(Func<DateTime> todaySupplier, DateTime birthday) : ICommand
{
    private readonly DateTime today = todaySupplier();

    public string ExecuteAndDisplayResult()
    {
        var daysBetween = (birthday - today).Days;
        if (daysBetween < 0)  // If birthday passed this year
        {
            // Calculate days until next year's birthday
            var nextBirthday = birthday.AddYears(1);
            daysBetween = (nextBirthday - today).Days;
        }
        return $"Still {daysBetween} days left!";
    }
}